### PR TITLE
[SPARK-40780][CONNECT] Add WHERE to Connect proto and DSL

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Add WHERE to Connect proto and DSL.

### Why are the changes needed?

Improve Connect proto testing coverage.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -70,7 +70,7 @@ object SimpleAnalyzer extends Analyzer(
     FakeV2SessionCatalog,
     new SessionCatalog(
       new InMemoryCatalog,
-      EmptyFunctionRegistry,
+      FunctionRegistry.builtin,
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:73]: Is this fix related?

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:73]: Yes. Without this line of change, we cannot use SimpleAnalyzer to resolve `UnresolvedFunction` because  this is `EmptyFunctionRegistry`.

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:73]: More specifically: we are testing `<` this time: `transform(connectTestRelation.where("id".protoAttr < 0))`, which becomes `UnresolvedFunction(name=<)`

Comment (grundprinzip) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:73]: Is this guaranteed to be only used in testing?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 2]
```diff
@@ -44,7 +45,20 @@ package object dsl {
     implicit class DslExpression(val expr: proto.Expression) {
       def as(alias: String): proto.Expression = proto.Expression.newBuilder().setAlias(
         proto.Expression.Alias.newBuilder().setName(alias).setExpr(expr)).build()
+
+      def < (other: proto.Expression): proto.Expression =
+        proto.Expression.newBuilder().setUnresolvedFunction(
+          proto.Expression.UnresolvedFunction.newBuilder()
+            .addParts("<")
```
Comment (grundprinzip) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:52]: does this actually work and we resolve '<' magically as an unresolved binary function?

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:52]: similarly would this work for `<=` or `=`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:73]: Yes

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala (hunk 2 above)
Comment (cloud-fan) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:52]: Yes
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L764-L771

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by HyukjinKwon — COMMENTED** [review]
Okie, LGTM. Should fix the docs at SimpleAnalyzer though.  let me make a quick PR.
